File: src/read_gmsh.jl
import PyPlot ; const plt = PyPlot
import PyCall: pyimport
mpl =pyimport("matplotlib")
mpl.use("Qt5Agg")

"""
Read a gmsh file (version 2) of 2D quadrangles.
Return an array of nodes positions and an array of the nodes composing each element.
"""
function readGmsh(meshFile)

    nodes = Array{Float64}[]
    elements = Array{Int}[]

    open(meshFile, "r") do f

        while ! eof(f)

            if readline(f) == "\$Nodes"
                s = readline(f) # numbre of nodes

                s = readline(f)
                while s != "\$EndNodes"
                    l = map(x->parse(Float64, x), split(s))
                    push!(nodes, l[2:3])
                    s = readline(f)
                end

                readline(f) ; readline(f); readline(f) # comments

                s = readline(f)
                while s != "\$EndElements" 
                    l = map(x->parse(Int64, x), split(s))
                    l[2] == 3 && push!(elements, l[6:9])
                    s = readline(f)
                end
            end
        end
    end

    return nodes, elements

end


function plotMesh(nodes, elements)

    fig, ax = plt.subplots()

    ax.scatter([n[1] for n in nodes], [n[2] for n in nodes], zorder=5)

    for el in elements

        n1,n2,n3,n4 = el
        n1,n2,n3,n4 = nodes[n1],nodes[n2],nodes[n3],nodes[n4]
        
        plt.plot([n1[1],n2[1]], [n1[2],n2[2]], "k")
        plt.plot([n2[1],n3[1]], [n2[2],n3[2]], "k")
        plt.plot([n3[1],n4[1]], [n3[2],n4[2]], "k")
        plt.plot([n4[1],n1[1]], [n4[2],n1[2]], "k")
    end

    ax.set_aspect("equal", "box")
    plt.show()
end

if abspath(PROGRAM_FILE) == @__FILE__ 

    nodes, elements = readGmsh("../gmsh_meshes/octogon.msh")
    println(length(nodes), " nodes")
    println(length(elements), " elements")
    plotMesh(nodes, elements)

end
nodes, elements = readGmsh("./gmsh_meshes/octogon.msh")
println(length(nodes), " nodes")
println(length(elements), " elements")
plotMesh(nodes, elements)

File: src/read_gmsh.py
import matplotlib.pyplot as plt

import numpy as np

def readGmsh(meshFile):
    """
    Read a gmsh file (version 2) of 2D quadrangles.
    Return an array of nodes positions and an array of the nodes composing each element.
    """

    nodes = []
    elements = []

    with open(meshFile, "r") as f:

        lines = f.readlines()
        nodesStart, nodesEnd, elemStart, elemEnd = 0,0,0,0

        for (i,l) in enumerate(lines):
            if l == "$Nodes\n": nodesStart = i+2
            elif l == "$EndNodes\n": nodesEnd = i
            elif l == "$Elements\n": elemStart = i+2
            elif l == "$EndElements\n": elemEnd = i

        print(nodesStart, nodesEnd, elemStart, elemEnd)

        nodes = [list(map(float, l.split()[1:3])) for l in lines[nodesStart:nodesEnd]]
        for l in lines[elemStart:elemEnd]:
            if len(l.split())==9:
                elements.append(list(map(int, l.split()[5:])))

    return np.array(nodes), np.array(elements)


def plotMesh(nodes, elements):

    fig, ax = plt.subplots()

    for e in elements:
        n1,n2,n3,n4 = e
        ne = [nodes[n1-1],nodes[n2-1],nodes[n3-1],nodes[n4-1]]
        plt.fill([n[0] for n in ne], [n[1] for n in ne], color="greenyellow", alpha=0.5, ec="k")

    ax.set_aspect("equal", "box")

    plt.show()


if __name__ == "__main__":

    nodes, elements = readGmsh("../gmsh_meshes/octogon.msh")
    print(f"{len(nodes)} nodes")
    print(f"{len(elements)} elements")
    plotMesh(nodes, elements)